Sources/BSON/BSONDocument.swift

```diff
@@ -174,12 +185,107 @@ public struct BSONDocument {
      */
     public subscript(dynamicMember member: String) -> BSON? {
         get { self[member] }
-        set { fatalError("Unimplemented") }
+        set { self[member] = newValue }
+    }
+
+    /**
+     * Sets a BSON element with the corresponding key
+     * if element.value is nil the element is deleted from the BSON
+     */
+    internal mutating func set(key: String, of value: BSON?) throws {
+        if !self.keySet.contains(key), let value = value {
+            // appending new key
+            self.keySet.insert(key)
+            self._buffer.moveWriterIndex(to: self.byteLength - 1) // setup to overwrite null terminator
+            let size = BSONDocument.append(element: (key, value), to: &self._buffer)
+            self._buffer.writeInteger(0, endianness: .little, as: UInt8.self) // add back in our null terminator
+            self.byteLength += size
+            return
+        }
+
+        guard value != nil || self.keySet.contains(key) else {
+            // no-op: trying to delete a key that doesn't exist
```

Comment:
can we add a test for this behavior? 

also just a nit but I think the logic might be clearer if we can combine this with the check above rather than checking whether the key is in the set and the value is non-nil twice, maybe something like

```
if keyset doesn't contain key:
    guard let value = value else return early (no op case)
    add new key
    return
```